src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+/*
+ *   Copyright (c) 2023 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+package matter.controller
+
+import chip.devicecontroller.ChipDeviceController
+import chip.devicecontroller.ChipDeviceControllerException
+import chip.devicecontroller.GetConnectedDeviceCallbackJni.GetConnectedDeviceCallback
+import chip.devicecontroller.InvokeCallback
+import chip.devicecontroller.ReportCallback
+import chip.devicecontroller.ResubscriptionAttemptCallback
+import chip.devicecontroller.SubscriptionEstablishedCallback
+import chip.devicecontroller.WriteAttributesCallback
+import chip.devicecontroller.model.AttributeWriteRequest
+import chip.devicecontroller.model.ChipAttributePath
+import chip.devicecontroller.model.ChipEventPath
+import chip.devicecontroller.model.ChipPathId
+import chip.devicecontroller.model.InvokeElement
+import java.util.logging.Level
+import java.util.logging.Logger
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlinx.coroutines.channels.Channel.Factory.UNLIMITED
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.onFailure
+import kotlinx.coroutines.channels.trySendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+/** Controller to interact with the CHIP device. */
+class MatterControllerImpl(params: ControllerParams) :
+  MatterController, chip.devicecontroller.ChipDeviceController.CompletionListener {
+  private val deviceController: ChipDeviceController
+  private var completionListener: MatterController.CompletionListener? = null
+  private var nodeId: Long? = null
+
+  override fun setCompletionListener(listener: MatterController.CompletionListener) {
+    completionListener = listener
```

Comment:
Is the only purpose of the listener to convert a `chip.devicecontroller.ChipDeviceController.CompletionListener` to a `MatterControllerListener` ? if MatterControllerImpl does not actually need access to the CompletetionListener, could we separate logic out in an adapter?

I would propose to have something like:

```kotlin
override fun setCompletionListener(listener: MatterController.CompletionListener?) =
    deviceController.setCompletionListener(CompletionListenerAdapter.from(listener));
```

asking for:
  - nullable (to support removing listeners if their lifetime is done)
  - remove the complexity from MatterControllerImpl

```kotlin
class CompletionListenerAdapter(val listener MatterController.CompletionListener) : 
        chip.devicecontroller.ChipDeviceController.CompletionListener {

   override fun onConnectDeviceComplete() = listener.onConnectDeviceComplete()
   override fun onStatusUpdate(status: Int) = listener.onStatusUpdate(status)
   // ...

    companion object {
        chip.devicecontroller.ChipDeviceController.CompletionListener? from(listener: MatterController.CompletionListener?){
            if (listener == null) {
                return null;
            }
            return CompletionListenerAdapter(listener);
        }
    }
}
```

Reply:
This is smart, thanks!